### PR TITLE
Update UX of authentication methods' Group Mappings Widget

### LINKED ISSUE #27212: FE - Make SSO group mapping clearer
This is an update to the SSO group mapping UX to make the semantics clearer and (hopefully) resolve some recurring confusions about how group mapping works. The original bug motivating this project was https://github.com/metabase/metabase/issues/26079

**Links**
- product doc: [Make SSO group mapping clearer](https://www.notion.so/metabase/Make-SSO-group-mapping-clearer-b9ed33e4ee8c4facb2220f20b565e51f)
- figma link: [link](https://www.figma.com/file/Q75JpYNz6D109ZjoTGZ0ZP/Make-SSO-group-mapping-clearer?node-id=1%3A4&t=FeMud61cDfdX077F-0)

**Implementation Plan**
https://github.com/metabase/metabase/pull/26555 is a backend PR to add a new endpoint for clearing the members of a specific group. Once this is merged, the frontend can implement the changes described in the product doc and use the new endpoint for the `Remove mapping and members` option.

The actual update should be performed at the end when users click "Save". This means we should accumulate and correctly update a list of groups to be cleared/removed while editing mappings including corner cases of re-adding the same mapping, etc.

### PR DESCRIPTION
**Pull request by gusaiani**
Fixes https://github.com/metabase/metabase/issues/27212

Makes the "Group Mappings Widget" inlined in 3 authentication method forms inside our Admin.

Adds ability to delete groups, or empty groups of users, along with deleting a mapping.

Updates UI details.

Specifications:
[Product Doc](https://www.notion.so/metabase/Make-SSO-group-mapping-clearer-b9ed33e4ee8c4facb2220f20b565e51f)

Here we also extract components away from the `GroupMappingsWidget` file, which was large.
We apply TypeScript to extracted files.
We add test coverage to the feature using jest.

### To verify

Use application in Enterprise mode.

The widget appears in these 3 pages:

http://localhost:3000/admin/settings/authentication/ldap
http://localhost:3000/admin/settings/authentication/saml
http://localhost:3000/admin/settings/authentication/jwt

Scroll to the bottom, as the widget lives near the end of the form.

Once there, you should see it:

image

You should be able to create a `New Mapping`, by giving it a unique name and hitting `Save`.
Once created, you can assign one or more groups to it.

Creating and editing a mapping happens async, that is, the changes are saved without submitting the form.

#### Deleting a mapping.

Once you've created a mapping, the delete flow can be slightly different depending on what groups if any are mapped:

**No group, or only Admin group:** a confirm modal will be shown. Clicking on `Yes` deletes the mapping.

**One or more groups besides Admin group:** a different modal will be shown. 
This will offer the possibility of:

1. Just delete the mapping. Groups will stay as they are.
2. Clear groups of their members. Groups will not be deleted, but all users in it will be removed from group.
3. Delete groups. Groups will be deleted. Users will *not* be removed from system.

To verify updates to groups after choosing option 2 or 3 above, visit http://localhost:3000/admin/people/groups.

#### Submitting the form — LDAP example

<details>
  <summary>To configure LDAP locally</summary>

[This PR](https://github.com/metabase/metabase/pull/25439) has good instructions.  

In short, you run a docker image: `docker run -p 389:389 osixia/openldap:1.5.0`

Then visit http://localhost:3000/admin/settings/authentication/ldap

Fill inputs with:

**LDAP Host:** localhost
**LDAP Port**: 389
**Username or DN:** cn=admin,dc=example,dc=org
**Password:** admin
**User search base:** dc=example,dc=org
</details>

### CODE REVIEW

**Comment by mazameli** [comment]
Couple small things I'm seeing that I'll add one by one.

First one is that I shouldn't see "No mappings yet" after I click New Mapping to add my first mapping:

image

**Comment by mazameli** [comment]
We'll want some padding to the left and right of this input so that it doesn't butt up right against the containing div.

image

File: frontend/src/metabase/admin/people/components/GroupSelect.jsx [hunk 1]
```diff
@@ -17,7 +17,7 @@ import GroupSummary from "./GroupSummary";
 
 export const GroupSelect = ({
   groups,
-  selectedGroupIds = new Set(),
```
Comment (gusaiani) [inline frontend/src/metabase/admin/people/components/GroupSelect.jsx:20]: Methods applied to `selectedGroupIds` down the line were always "includes".

We have been passing arrays to this props.

Once TypeScript reached a parent component of this one, the inconsistency beeped.

**Comment by mazameli** [comment]
Maybe this is because I put nonsense into my nonexistant LDAP settings here, but I didn't get any onscreen feedback after turning this toggle on that anything was wrong, and then when I refreshed the page the toggle was turned back off:

toggle-not-saved

**Comment by gusaiani** [comment]
@mazameli 

> but I didn't get any onscreen feedback after turning this toggle on that anything was wrong, and then when I refreshed the page the toggle was turned back off:

Thanks.
Yes, I've been learning that these authentication forms seem to have several things that deserve improvement.

<details>
  <summary>To configure LDAP locally</summary>

[This PR](https://github.com/metabase/metabase/pull/25439) has good instructions.  

In short, you run a docker image: `docker run -p 389:389 osixia/openldap:1.5.0`

Then visit http://localhost:3000/admin/settings/authentication/ldap

Fill inputs with:

**LDAP Host:** localhost
**LDAP Port**: 389
**Username or DN:** cn=admin,dc=example,dc=org
**Password:** admin
**User search base:** dc=example,dc=org
</details>

Atm in my mind a good approach would be:

1. to see if the groups widget itself is satisfactory working independently from submitting the whole form — this is just about the groups themselves, as "enable/disable" currently go with the form.
2. file separate issues about form validation and UX of submitting the form 
3. file an issue about the message returned by the server when adding a group whose name is not valid (current message could be made more helpful…telling us: what names are ok?)

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.unit.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+      userEvent.click(screen.getByLabelText(/Also remove all group members/i));
+      userEvent.click(
+        await screen.findByRole("button", {
+          name: "Remove mapping and members",
+        }),
+      );
+
+      expect(await screen.findByText("No mappings yet")).toBeInTheDocument();
+
+      expect(nock.isDone()).toBeTruthy();
```
Comment (gusaiani) [inline frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.unit.spec.tsx:144]: Here I want to check that the endpoints hit when deleting a mapping (and choosing to clear associated groups of members).

Is there a better way to check this?

**Comment by alxnddr** [comment]
Hey @gusaiani, for some reason JWT auth page crashes:

<video src="https://user-images.githubusercontent.com/14301985/218867015-770640a4-9933-4c07-aa57-6cf6ba804ff9.mov" />

**Comment by gusaiani** [comment]
> Hey @gusaiani, for some reason JWT auth page crashes

Thanks @alxnddr. What error message are you getting in the console?

**Comment by alxnddr** [comment]
It is possible to create an invalid mapping which will make the page crash without the ability to fix it from a user perspective.

Steps to reproduce:
1) Create a group "DELETE_ME"
2) Settings -> Auth -> SAML -> Create mapping "1" with "DELETE_ME" group.
3) Remove the mapping and select that you want to delete groups too
4) Create a mapping "2", open the group picker, `DELETE_ME` is still there. Create a mapping with it.

Now the page is broken. I believe we need to add:
- BE logic that removes all mappings related to deleted groups
- FE `componentDidCatch` on a mapping row with a delete button that will allow to remove a corrupted mapping

**Review by mazameli — APPROVED** [review]
I appreciate that you called out that there's some weirdness with how you have to hit Save to make the toggle's state stick, but that the adding or removing of groups doesn't require clicking Save. I compared this behavior to how it currently works in `master` on our Stats instance, and it's effectively the same thing, modal aside. I could imagine a hypothetical where this could create problems for people in case they weren't ready for a change to take effect for example, but again, this is no different from the status quo. The inline UX is the real gain here, and I don't think we should block that on needing to rewire things.

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.jsx [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+  useEffect(() => {
+    async function fetchData() {
+      const setting = _.findWhere(await SettingsApi.list(), {
+        key: mappingSetting,
+      });
+
+      setMappings(setting?.value || {});
+
+      PermissionsApi.groups().then(groups =>
```
Comment (alxnddr) [inline frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.jsx:40]: nit: mix of `await` and `then` in this function

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.jsx [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+function GroupMappingsWidget({ mappingSetting, ...props }) {
+  const [showAddRow, setShowAddRow] = useState(false);
+  const [groups, setGroups] = useState(null);
+  const [mappings, setMappings] = useState({});
+  const [saveError, setSaveError] = useState(null);
+
+  useEffect(() => {
+    async function fetchData() {
+      const setting = _.findWhere(await SettingsApi.list(), {
```
Comment (alxnddr) [inline frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.jsx:34]: Fetching data here can be simplified with https://streamich.github.io/react-use/?path=/story/side-effects-useasync--docs

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.unit.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    });
+
+    it("adds mapping to table on success", async () => {
+      setup();
+
+      userEvent.click(screen.getByText("New mapping"));
+      const input = screen.getByLabelText("new-group-mapping-name-input");
+      userEvent.type(input, "cn=Group");
+      act(() => {
+        screen.getByRole("button", { name: "Add" }).click();
```
Comment (alxnddr) [inline frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.unit.spec.tsx:45]: It would be nice to use `userEvent` here as well

**Review by alxnddr — COMMENTED** [review]
Looks great 🏅 

As we discussed in Slack, it would be great to create a follow-up PR that uses actions for removing/clearing groups to get the app state updated. But it would be an improvement to what we had before so it should not block merging.

It also would be really great to add e2e specs at least on one usage of the widget since unit specs mock a lot

**Comment by gusaiani** [comment]
@alxnddr, thanks very much for catching the error around deleted groups.

> Now the page is broken. I believe we need to add:

> BE logic that removes all mappings related to deleted groups

BE is already smart in that, when loading mappings, if a group has been deleted, it will remove it from a mapping.

I've updated the PR so that the FE now handles this group state update.

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.unit.spec.tsx (hunk 5 above)
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.unit.spec.tsx]: Oh yes, done 👍

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.jsx (hunk 3 above)
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/GroupMappingsWidget.jsx]: 👍 updated

**Comment by alxnddr** [comment]
The page crashes for me:
<video src="https://user-images.githubusercontent.com/14301985/219439451-8a9cbf95-e7c0-49cd-a9bb-0d715d64057d.mov" />

Also, it would be a really nice follow-up fix:
- While creating a mapping pressing an `Enter` does not save the mapping but cancels the input
- Error is rendered on the right, it would be nice to at least move it to the bottom

Screen Shot 2023-02-16 at 2 11 21 PM

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      <div>
+        <ModalHeader>{t`Remove this group mapping?`}</ModalHeader>
+        <ModalSubtitle>{subtitle}</ModalSubtitle>
+        <ModalRadioRoot>
+          <p>{whatShouldHappenText}</p>
+
+          <Radio
+            className="ml2"
+            vertical
+            value={value as DeleteMappingModalValueType | undefined}
```
Comment (alxnddr) [inline frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx:73]: Seems like we can omit this typecasting

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+              },
+            ]}
+            showButtons
+            onChange={handleChange}
+          />
+        </ModalRadioRoot>
+        <ModalFooter fullPageModal={false} formModal={true}>
+          <Button onClick={onHide}>{t`Cancel`}</Button>
+          <Button danger onClick={handleConfirm}>
+            {submitButtonLabels[value as DeleteMappingModalValueType]}
```
Comment (alxnddr) [inline frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx:98]: Unnecessary typecasting

**Review by alxnddr — COMMENTED** [review]
Looks good! Please fix the page crash and it will be good to merge

File: frontend/test/metabase/scenarios/admin/settings/sso/ldap.cy.spec.js [hunk 8]
```diff
@@ -106,6 +106,61 @@ describe(
...
+
+        cy.visit("/admin/people/groups");
+        cy.findByText("data").should("not.exist");
+        cy.findByText("nosql").should("not.exist");
+
+        checkThatGroupHasNoMembers("collection");
+        checkThatGroupHasNoMembers("readonly");
+      });
+
+      it("should allow deleting mappings with groups, while keeping remaining mappings consistent with their undeleted groups", () => {
```
Comment (alxnddr) [inline frontend/test/metabase/scenarios/admin/settings/sso/ldap.cy.spec.js:138]: 👍

**Review by alxnddr — APPROVED** [review]
Thanks for fixes and more specs 👏
